doc/conf.py

```diff
@@ -142,41 +139,43 @@
 # a list of builtin themes.
 html_theme = 'pydata_sphinx_theme'
 
-# Tell the theme where the code lives
-
-
-def _custom_edit_url(github_user, github_repo, github_version, doc_path, file_name, default_edit_page_url_template):
-    """Create custom 'edit' URLs for API modules since they are dynamically generated."""
-    if file_name.startswith("api/"):
-        # this is a dynamically generated API page, link to actual Python source
-        modpath = os.sep.join(os.path.splitext(file_name[4:])[0].split("."))
-        if modpath == "modules":
-            # main package listing
-            modpath = "vispy"
-        rel_modpath = os.path.join("..", modpath)
-        if os.path.isdir(rel_modpath):
-            doc_path = modpath + "/"
-            file_name = "__init__.py"
-        elif os.path.isfile(rel_modpath + ".py"):
-            doc_path = os.path.dirname(modpath)
-            file_name = os.path.basename(modpath) + ".py"
-        else:
-            warnings.warn(f"Not sure how to generate the API URL for: {file_name}")
-    return default_edit_page_url_template.format(github_user=github_user,
-                                                 github_repo=github_repo,
-                                                 github_version=github_version,
-                                                 doc_path=doc_path,
-                                                 file_name=file_name)
-
+# Create custom 'edit' URLs for API modules since they are dynamically generated.
+# We precompute this so the values in the `html_context` are static, and it can be cached
+# `modules.rst` is a special case, and we link it to the main `vispy` package
+edit_link_paths = {"api/modules.rst": "vispy/__init__.py"}
+for root, dirs, files in os.walk("../vispy"):
+    # remove leading "../"
+    root = root[3:]
+    if root.endswith("__pycache__"):
+        continue
+    for file in files:
+        full_path = os.path.join(root, file)
+        if full_path.endswith("__init__.py"):
+            package_name = root.replace(os.sep, ".")
+            apidoc_file_name = "api/" + package_name + ".rst"
+        elif full_path.endswith(".py"):
+            module_name = os.path.splitext(full_path)[0].replace(os.sep, ".")
+            apidoc_file_name = "api/" + module_name + ".rst"
+        edit_link_paths[apidoc_file_name] = full_path
```

Comment:
Clever. Not bad. My one small request (as the person not doing any of this work so feel free to ignore it) is that this be done with Pathlib for maybe cleaner code...maybe.